lib/compiler/src/sys_core_alias.erl

```diff
@@ -0,0 +1,317 @@
+    Sub2 = sub_unfold(Sub1#sub{v=V,t=T}),
+    {Node#c_clause{pats=Pats1},Sub2};
+
+post(Node, Sub) ->
+    {Node,Sub}.
+
+%% sub_new/1
+%% sub_add_keys/2
+%% sub_take_keys/3
+%% sub_cache_nodes/3
+%%
+%% Manages the substitutions record.
+
+%% Builds a new sub.
+-spec sub_new(term()) -> sub().
+sub_new(Temp) ->
+    #sub{t=Temp}.
+
+%% Folds the sub into a new one if the variables in nodes are not disjoint
+sub_fold(VarNames, #sub{v=SubNames}=Sub) ->
+    case is_disjoint_variables(VarNames, SubNames) of
+	true  -> Sub#sub{t={temp,Sub#sub.t}};
+	false -> sub_new({sub,Sub})
+    end.
+
+%% Unfolds the sub in case one was folded in the previous step
+sub_unfold(#sub{t={temp,Temp}}=Sub) ->
+    Sub#sub{t=Temp};
+sub_unfold(#sub{t={sub,Sub}}) ->
+    Sub.
+
+%% Adds the keys extracted from patterns to the state.
+-spec sub_add_keys([term()], sub()) -> {sub(),[term()]}.
+sub_add_keys(Keys0, #sub{p=Pat0}=Sub) ->
+    {Pat1,Keys1} = sub_add_keys(Keys0, Pat0, []),
+    {Sub#sub{p=Pat1},Keys1}.
+
+sub_add_keys([H|T], Pat, Keys) ->
+    case Pat of
+	#{H := _} ->
+	    %% We have already seen the pattern upstream
+	    sub_add_keys(T, Pat, Keys);
+	#{} ->
+	    sub_add_keys(T, maps:put(H, ?NOTSET, Pat), [H|Keys])
+    end;
+sub_add_keys([], Pat, Keys) ->
+    {Pat, lists:reverse(Keys)}.
+
+%% Take the keys from the map taking into account the keys
+%% that have changed as those must become aliases in the pattern.
+-spec sub_take_keys([term()], sub()) -> {sub(), [{term(), atom()}]}.
+sub_take_keys(Keys, #sub{p=Pat0}=Sub) ->
+    {Pat1,Acc} = sub_take_keys(Keys, Pat0, []),
+    {Sub#sub{p=Pat1},Acc}.
+
+sub_take_keys([K|T], Sub0, Acc) ->
+    case maps:take(K, Sub0) of
+	{?NOTSET,Sub1} ->
+	    sub_take_keys(T, Sub1, Acc);
+	{Name,Sub1} ->
+	    sub_take_keys(T, Sub1, [{K,Name}|Acc])
+    end;
+sub_take_keys([], Sub, Acc) ->
+    {Sub,Acc}.
+
+%% Check if the node can be cached based on the state information.
+%% If it can be cached and it does not have an alias for it, we
+%% build one.
+-spec sub_cache_nodes(atom(), [cerl:cerl()], sub()) -> {atom(), sub()} | error.
+sub_cache_nodes(Kind, Nodes, #sub{p=Pat}=Sub) ->
+    case nodes_to_key(Kind, Nodes) of
+	{ok, Key} ->
+	    case Pat of
+		#{Key := ?NOTSET} ->
+		    new_var_name(Key, Sub);
+		#{Key := Name} ->
+		    {Name,Sub};
+		#{} ->
+		    error
+	    end;
+	error ->
+	    error
+    end.
+
+new_var_name(Key, #sub{p=Pat}=Sub) ->
+    Counter = get(new_var_num),
+    Name = list_to_atom("@r" ++ integer_to_list(Counter)),
+    put(new_var_num, Counter + 1),
+    {Name,Sub#sub{p=maps:put(Key, Name, Pat)}}.
+
+%% get_variables/1
+%% is_disjoint_variables/2
+%% merge_variables/2
+
+get_variables(NodesList) ->
+    lists:foldl(fun(Node, Acc) ->
+	ordsets:union(cerl_trees:variables(Node), Acc)
+    end, [], NodesList).
+
+is_disjoint_variables(Vars1, Vars2) ->
+    ordsets:is_disjoint(Vars1, Vars2).
+
+merge_variables(Vars1, Vars2) ->
+    ordsets:union(Vars1, Vars2).
```

Comment:
That could be a good idea. Running eprof shows that `ordsets:union/2` does seem to consume a significant part of the run time:

```
$ erlc '+{eprof,sys_core_alias}' -S core_parse.erl
.
.
.
lists:foldl/3                                           42182     4.15    7123  [      0.17]
sys_core_alias:'-def/1-anonymous-0-'/2                  39306     6.10   10476  [      0.27]
cerl:is_data/1                                          38458     6.27   10762  [      0.28]
cerl_trees:mapfold_list/4                               34582     7.87   13515  [      0.39]
cerl_trees:mapfold/4                                    39306     8.78   15066  [      0.38]
ordsets:union/2                                         70616    10.19   17492  [      0.25]
cerl:type/1                                             62350    10.62   18225  [      0.29]
-----------------------------------------------------  ------  -------  ------  [----------]
Total:                                                 695127  100.00%  171679  [      0.25]

```